File: src/lib/health-id.ts
import { prisma } from "@/lib/prisma";

export function generateHealthId(
  name: string,
  phoneNumber: string,
  gender: "male" | "female" | "other"
): string {
  const nameHash = name
    .toLowerCase()
    .split("")
    .reduce((a, b) => a + b.charCodeAt(0), 0);
  const phoneHash = phoneNumber
    .split("")
    .reduce((a, b) => a + b.charCodeAt(0), 0);
  const genderCode = gender === "male" ? "1" : gender === "female" ? "2" : "3";
  const uniqueId = (nameHash + phoneHash).toString().padStart(5, "0").slice(0, 5);
  const random = Math.floor(Math.random() * 100)
    .toString()
    .padStart(2, "0");
  const today = new Date();
  const year = today.getFullYear().toString();
  const month = (today.getMonth() + 1).toString().padStart(2, "0");
  return `${year}${month}${uniqueId}${genderCode}${random}`;
}

export async function ensureUniqueHealthId(
  name: string,
  phoneNumber: string,
  gender: "male" | "female" | "other"
): Promise<string> {
  const existing = await prisma.user.findFirst({
    where: {
      type: "patient",
      name: { equals: name },
      phoneNumber,
    },
  });
  if (existing?.healthId) {
    return existing.healthId;
  }

  let healthId = generateHealthId(name, phoneNumber, gender);
  let attempts = 0;
  while (await prisma.user.findUnique({ where: { healthId } })) {
    healthId = generateHealthId(name, phoneNumber, gender);
    attempts += 1;
    if (attempts > 20) {
      throw new Error("Could not generate a unique Health ID");
    }
  }
  return healthId;
}

File: backend/app/services/health_id.py
import random
from datetime import datetime

from sqlalchemy.orm import Session

from app.db.models import User


def generate_health_id(name: str, phone_number: str, gender: str) -> str:
    name_hash = sum(ord(c) for c in name.lower())
    phone_hash = sum(ord(c) for c in phone_number)
    gender_code = "1" if gender == "male" else "2" if gender == "female" else "3"
    unique_id = str(name_hash + phone_hash).zfill(5)[:5]
    rand = str(random.randint(0, 99)).zfill(2)
    today = datetime.now()
    return f"{today.year}{today.month:02d}{unique_id}{gender_code}{rand}"


def ensure_unique_health_id(
    db: Session, name: str, phone_number: str, gender: str
) -> str:
    existing = (
        db.query(User)
        .filter(
            User.type == "patient",
            User.name == name,
            User.phone_number == phone_number,
        )
        .first()
    )
    if existing and existing.health_id:
        return existing.health_id

    health_id = generate_health_id(name, phone_number, gender)
    for _ in range(20):
        if not db.query(User).filter(User.health_id == health_id).first():
            return health_id
        health_id = generate_health_id(name, phone_number, gender)
    raise RuntimeError("Could not generate a unique Health ID")
